APA102.class.nut

```diff
@@ -0,0 +1,169 @@
+// MIT License
+//
+// Copyright (c) 2016-17 Electric Imp
+//
+// SPDX-License-Identifier: MIT
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+const FRAME_LENGTH = 4;
+
+class APA102 {
+
+    static VERSION = "1.1.0";
+
+    // _spi should be pre-configured with SIMPLEX_TX and any clock speed
+    _spi = null;
+
+    // _clk and _dat are imp GPIO pins for APA102s connected via GPIO
+    _clk = null;
+    _dat = null;
+
+    // _mode is true for SPI usage; false for bit-bang mode
+    _mode = true;
+
+    // _num is the number of pixels in the chain
+    _num = 0;
+
+    // _data is a blob where every four bytes is frame controlling one LED
+    // plus head and tail frames for APA102 signalling. The head frame is
+    // four bytes of 0x00, the tail is four bytes of 0xFF (see constructor)
+    _data = null;
+
+    function constructor(spiBus = null, numPixels = 0, clockPin = null, dataPin = null) {
+        // To select your mode, pass in a valid imp SPI bus OR valid imp GPIO pins (clock and data)
+        // SPI mode is the default (_mode is true)
+        if (numPixels == 0) {
+            server.error("APA102() requires a non-zero number of pixels.");
```

Comment:
FWIW, the reason I don't generally throw is because I'm so used to checking for a returned null, but I see your point. I think throw for errors that should halt execution (eg. trying to call the constructor without passing in a required object) is a reasonable case to use throw, so I'll update accordingly.